### ISSUE TITLE
Q: How to handle sliding expiration in MVC correctly?

### ISSUE THREAD

**Issue by urbanhusky** [IdentityServer/IdentityServer3#2192]
Hi,

I've got an ASP.NET MVC application with a WebApi hosted on the same app. I'm using MVC to render the html, provide I18N etc., the actual client app uses angularjs to access the WebApi. The application was previously set up to use forms authentication.

I've disabled all authentication in the config and am using the cookie authentication OWIN middleware `UseCookieAuthentication`. For authentication against Ids I'm using resource owner password credentials flow on the login action where the forms authentication was previously done. Afterwards I read the claims from the userinfo endpoint, create a `ClaimsIdentity` and store that in the cookie.

Login and logout works fine so far. Validating the token and performing the authorization is another issue, which might be dependent on how I'll handle sliding expiration.

I want to log out inactive users after a brief period of time, but do not want to log out active users after the same time span (i.e. I need sliding expiration of their access to the entire app).

Q: How can I correctly implement a sliding expiration in this scenario? Would it be enough to set sliding expiration on the cookie and set a high enough access token lifetime? Currently the access token is not provided to the browser, but stored in the claims, encrypted in the cookie. (With the potential issue of CSRF vulnerability?)

**Comment by mderriey** [comment]
Hi,

Before going further, I would like some details on your setup.
When you say that the actual client application use angular to access the API, how is authorization done at the API level?

I'm asking because you say that the access token is stored in the cookie, so there's no way for the angular app to access the cookie, neither to send it to the API.
A typical setup would be that the access token is sent in the request to the API, so the authorization is done against the token. Here it looks like it's done against the cookie, maybe?

Cheers.

**Comment by urbanhusky** [comment]
Currently there is no authorization done aside from having cookie authentication (haven't checked if the ajax requests send the cookie, but I think they do, from how I experienced the application behaviour). We might add the token to all API calls later, right now I want to get this set up for a prototype asap.

**Comment by mderriey** [comment]
OK.
So it has nothing to do with IdentityServer specifically, more with OWIN cookie authentication handler.
I guess what you're after is a combination of the [`ExpireTimeSpan`](https://msdn.microsoft.com/en-us/library/microsoft.owin.security.cookies.cookieauthenticationoptions.expiretimespan(v=vs.113).aspx) and the [`SlidingExpiration`](https://msdn.microsoft.com/en-us/library/microsoft.owin.security.cookies.cookieauthenticationoptions.slidingexpiration(v=vs.113).aspx) properties.
The first one defines how long the cookie will be valid, and the second instructs the cookie middleware to issue a new cookie whenever the current cookie is past half of its lifetime.

**Comment by urbanhusky** [comment]
So it would be enough to use sliding expiration on the cookie with a long lived access token?

**Comment by mderriey** [comment]
I'd say so.
Once again, not something I'd recommend, but it should work since both the cookie and the token will be valid.

**Comment by urbanhusky** [comment]
What would you recommend? I mean I fully intend to switch the API over to bearer token authentication but that's a bit tricky right now since we're hosting both the MVC and API in the same web application, we didn't start out with OWIN etc. - so setting up auth for the API separately is currently a bit hard.

**Comment by mderriey** [comment]
I'd recommend doing authorization against a token, but I get that it's difficult for you to achieve.
We can have this conversation later when you're happy with the prototype and willing to go forward

**Comment by urbanhusky** [comment]
For authorization, I'm planning on doing claims-based authorization, using the claims which I get from the token.

**Comment by sweetsruthi** [comment]
Hi I have same issue. Can you let me know how to extend cookie expire time when user is working on ajax calls?
In my application, when user working a page which has ajax calls. it is logging him out after one hour eventhough the user is active.

**Comment by urbanhusky** [comment]
I'm not exactly sure how I solved it back then, but I do remember that it was somewhat ugly (well, I'm not too happy about the entire project in retrospect). I can't look up how I solved it, but I thing I had a service which either polled or used a websocket/signalr to inform the app that the cookie needs to be refreshed.

Basically it worked like this:
API call -> check remaining cookie lifetime -> if less than half remaining, set flag _cookie refresh required_ -> inform app -> app calls MVC controller, which causes the cookie to refresh.

This however meant that there is an endpoint which anybody could call to extend the lifetime of a cookie. We might have used some sort of auth mechanism (providing a secret, e.g. a hash which corresponds to allowing the app to refresh a cookie with a specific remaining lifetime, once).

All in all, this was way too complicated imho and I wouldn't exactly recommend doing this.
